Update tsconfig.json

colindjk opened the pull request:
Add support for preact-cli `async!` prefix.

Referenced here - https://github.com/preactjs/preact-cli/blob/1.4.1/examples/full/src/components/app.js#L7.

rschristian commented on template/tsconfig.json:
```diff
@@ -35,7 +35,9 @@
         // "baseUrl": "./",                       /* Base directory to resolve non-absolute module names. */
-        // "paths": {},                           /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
+        "paths": {                                /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
+            "async\!*": ["*"]
```
The escape sequence is illegal.

colindjk replied:
I removed it.

rschristian commented:
This won't actually work in a correctly set up project. You'll just end up with: 

```
TS2307: Cannot find module 'async!./routes/home' or its corresponding type declarations.
```

colindjk commented:
> This won't actually work in a correctly set up project. You'll just end up with:
> 
> ```
> TS2307: Cannot find module 'async!./routes/home' or its corresponding type declarations.
> ```

The `paths` setting does not cover relative imports, I should've specified this originally. If you set `baseUrl` to your source directory it will work for absolute imports. 

AFAIK there is no way to have a relative import mapped to a different import, which means that the `async!` prefix would only be compatible with absolute imports. I'll add a comment next to the `paths` setting specifying this.

rschristian commented:
@colindjk I get that, but you can't just start making tsconfig paths as Webpack will not acknowledge them at build time.

colindjk commented:
I'm confused, preact-cli has support for imports with the `async!` prefix (webpack issue). TypeScript doesn't like this prefix unless you map it via paths. Ideally only component files would be able to have the `async!` prefix allowed, but that's not really possible from a tsconfig.

Is there another way of handling this?

rschristian commented:
Preact-CLI supports `async!`, yes, but it does not (out of the box) support TSConfig paths as Webpack does not. 

Furthermore, adding `async!` like that will not end up working due to the resolution. Say I have a tsconfig as follows: 

```
"baseUrl": ".",
"paths": {
   "async!/*": ["*"]
},
```

If I tried to use `import Home from "async!/src/routes/home";` I'd be met with: 

```
✖ ERROR ./components/app.tsx
Module not found: Error: Can't resolve '/src/routes/home' in './src/components
```

As far as I know it can't be done, someone was trying a few weeks ago and couldn't get it set up correctly, IMO, this is exactly the type of JS weirdness you'd want to leave behind when using TS anyways, but that's just my opinion. 

The `routes/` and `async/` directories are automatically code split if that'll work for you.

colindjk commented:
> Preact-CLI supports `async!`, yes, but it does not (out of the box) support TSConfig paths as Webpack does not.
> 
> Furthermore, adding `async!` like that will not end up working due to the resolution. Say I have a tsconfig as follows:
> 
> ```
> "baseUrl": ".",
> "paths": {
>    "async!/*": ["*"]
> },
> ```
> 
> If I tried to use `import Home from "async!/src/routes/home";` I'd be met with:
> 
> ```
> ✖ ERROR ./components/app.tsx
> Module not found: Error: Can't resolve '/src/routes/home' in './src/components
> ```
> 
> As far as I know it can't be done, someone was trying a few weeks ago and couldn't get it set up correctly, IMO, this is exactly the type of JS weirdness you'd want to leave behind when using TS anyways, but that's just my opinion.
> 
> The `routes/` and `async/` directories are automatically code split if that'll work for you.

In that case, just remove the forward slash you have in `async!` to get it work! 

Given that the support for `async!` is this contentious, could there at least be something added to the README specifying that this functionality isn't supported in the same way?

Thanks for taking the time to respond my queries.

rschristian commented:
> 
> In that case, just remove the forward slash you have in `async!` to get it work!

Which forward slash are you referring to? I'd love to see this config if you have it, as I've certainly played with the imports and nothing has passed `preact build` yet. You can probably satisfy your editor easily, but that's not going to work when it comes to building.
 
> Given that the support for `async!` is this contentious, could there at least be something added to the README specifying that this functionality isn't supported in the same way?

I'm not sure I'd say it's contentious. I certainly disagree with it's use, as I disagree with a couple of the TS related things, but that's just my opinion. You can just `@ts-ignore` your way through the problem.

Feel free to make a PR.

colindjk commented:
@rschristian I realized there was another line to the preact config that I had added, sorry for wasting your time!

I've added it now (resolves absolute paths to `src/`). 

At this point though, the change is more than I was hoping it would be / it's a bit more opinionated :/. I can close this PR and make another one for the README so people are aware the `async!` import syntax isn't supported.

rschristian commented:
Ah, gotcha, that makes a lot more sense. 

I do think at this point this is too opinionated as it doesn't match the other templates well. And I do set that very same config on every one of my own projects, mind you.

I'm wondering if it's absolutely impossible to deal with this in the CLI. I'll go make a note of it, but if you wanted to make a PR to update the CLI's readme with regards to `async!` it'd be much appreciated.

colindjk commented:
@rschristian Opened up https://github.com/preactjs/preact-cli/pull/1505, closing this PR now. Thanks for the patience + help!
